### PR TITLE
[A2-798] OPA update timing out -- part one: do less work

### PR DESCRIPTION
**Pull request by srenatus**
### :nut_and_bolt: Description

The removal of the extra branches saves us some time on the v2 init partial result, see the new benchmark:
```
name                                                 old time/op    new time/op    delta
InitPartialResultV2/default_policies,_run_0_times-8     1.68s ±20%     1.32s ± 8%  -21.71%  (p=0.000 n=10+10)
InitPartialResultV2/default_policies,_run_1_times-8     2.78s ± 4%     2.30s ± 3%  -17.50%  (p=0.000 n=9+10)
InitPartialResultV2/default_policies,_run_2_times-8     4.17s ± 5%     3.97s ±19%     ~     (p=0.211 n=9+10)
InitPartialResultV2/default_policies,_run_3_times-8     5.24s ± 2%     4.61s ± 6%  -12.02%  (p=0.000 n=10+10)
InitPartialResultV2/default_policies,_run_4_times-8     6.43s ± 1%     5.73s ± 6%  -10.97%  (p=0.000 n=9+10)
InitPartialResultV2/0_dummy_policies-8                  312ms ± 5%     292ms ± 4%   -6.21%  (p=0.000 n=10+9)
InitPartialResultV2/25_dummy_policies-8                 332ms ± 1%     346ms ±12%     ~     (p=0.408 n=10+8)
InitPartialResultV2/50_dummy_policies-8                 363ms ± 1%     345ms ± 3%   -4.90%  (p=0.000 n=10+8)
InitPartialResultV2/75_dummy_policies-8                 398ms ± 1%     456ms ±16%  +14.43%  (p=0.000 n=10+10)
InitPartialResultV2/100_dummy_policies-8                446ms ± 4%     493ms ±24%  +10.68%  (p=0.029 n=10+10)

name                                                 old alloc/op   new alloc/op   delta
InitPartialResultV2/default_policies,_run_0_times-8     307MB ± 0%     256MB ± 0%  -16.64%  (p=0.000 n=10+10)
InitPartialResultV2/default_policies,_run_1_times-8     564MB ± 0%     465MB ± 0%  -17.49%  (p=0.000 n=10+10)
InitPartialResultV2/default_policies,_run_2_times-8     821MB ± 0%     674MB ± 0%  -17.81%  (p=0.000 n=10+10)
InitPartialResultV2/default_policies,_run_3_times-8    1.08GB ± 0%    0.88GB ± 0%  -17.98%  (p=0.000 n=10+10)
InitPartialResultV2/default_policies,_run_4_times-8    1.33GB ± 0%    1.09GB ± 0%  -18.08%  (p=0.000 n=10+10)
InitPartialResultV2/0_dummy_policies-8                 67.9MB ± 0%    63.1MB ± 0%   -7.09%  (p=0.000 n=10+10)
InitPartialResultV2/25_dummy_policies-8                72.3MB ± 0%    67.4MB ± 0%   -6.66%  (p=0.000 n=10+10)
InitPartialResultV2/50_dummy_policies-8                77.7MB ± 0%    72.9MB ± 0%   -6.19%  (p=0.000 n=10+10)
InitPartialResultV2/75_dummy_policies-8                84.2MB ± 0%    79.4MB ± 0%   -5.71%  (p=0.000 n=10+10)
InitPartialResultV2/100_dummy_policies-8               91.7MB ± 0%    86.9MB ± 0%   -5.24%  (p=0.000 n=9+10)

name                                                 old allocs/op  new allocs/op  delta
InitPartialResultV2/default_policies,_run_0_times-8     10.2M ± 0%      8.6M ± 0%  -15.84%  (p=0.000 n=8+10)
InitPartialResultV2/default_policies,_run_1_times-8     18.6M ± 0%     15.5M ± 0%  -16.65%  (p=0.000 n=10+10)
InitPartialResultV2/default_policies,_run_2_times-8     27.0M ± 0%     22.4M ± 0%  -16.95%  (p=0.000 n=10+10)
InitPartialResultV2/default_policies,_run_3_times-8     35.4M ± 0%     29.4M ± 0%  -17.10%  (p=0.000 n=10+10)
InitPartialResultV2/default_policies,_run_4_times-8     43.8M ± 0%     36.3M ± 0%  -17.21%  (p=0.000 n=10+10)
InitPartialResultV2/0_dummy_policies-8                  2.52M ± 0%     2.33M ± 0%   -7.39%  (p=0.000 n=10+10)
InitPartialResultV2/25_dummy_policies-8                 2.67M ± 0%     2.48M ± 0%   -6.98%  (p=0.000 n=10+9)
InitPartialResultV2/50_dummy_policies-8                 2.85M ± 0%     2.66M ± 0%   -6.55%  (p=0.000 n=10+10)
InitPartialResultV2/75_dummy_policies-8                 3.05M ± 0%     2.87M ± 0%   -6.10%  (p=0.000 n=9+10)
InitPartialResultV2/100_dummy_policies-8                3.29M ± 0%     3.11M ± 0%   -5.66%  (p=0.000 n=10+10)
```
old is `automate@master`, new is this branch. Besides the improvements from this branch, this also proves that out issue is *one of state* not of **scale**: It doesn't matter (much) if there are 0 or 100 policies, but it matters **a lot** if you run `initPartialResultsV2` multiple times. That's what happens in a real-world system: for every update, it'll be re-run. And it'll take longer and longer and longer with each of them.

Taking a stab at fixing that is up next on my list for *A2-798*.

### :+1: Definition of Done

No functional changes 😉 

### :athletic_shoe: Demo Script / Repro Steps

NA

### :chains: Related Resources

- [A2-767](https://chefio.atlassian.net/browse/A2-767)
- [A2-798](https://chefio.atlassian.net/browse/A2-798)

### :white_check_mark: Checklist

- [X] Necessary tests added/updated?
- [ ] Necessary docs added/updated?
- [X] Code actually executed?
- [X] Vetting performed (unit tests, lint, etc.)?

### CODE REVIEW

File: components/authz-service/engine/opa/policy/authz_v2.rego [hunk 1]
```diff
@@ -102,14 +92,7 @@ denied_project[project] {
...
 }
 
-# when no input is provided, this rule short-circuits checking any other allowed projects
-denied_all_projects {
-	match[["deny", pol_id, statement_id]]
-	common.const_all_projects == policies[pol_id].statements[statement_id].projects[_]
-}
-
 authorized_project[project] {
-	not denied_all_projects
```
Comment (msorens) [inline components/authz-service/engine/opa/policy/authz_v2.rego]: I was going to say, "Wait, wait; this is needed for a special case!"  But not anymore. 😁

**Comment by srenatus** [comment]
ℹ️ So, to be clear: this doesn't solve the issue; it's a byproduct of an ongoing investigation (hence the PR rename). Nonetheless, I'll merge this and keep moving this forward off of master in another PR.

**Comment by srenatus** [comment]
🤦‍♂ the benchmark I've added runs on an empty store 😄 willfix.

**Comment by srenatus** [comment]
ℹ️ What I've changed over the last review is not that significant, and the real work for fixing the issue identified here will happen later -- I'll thus merge this when it's green.

**Pull request merged**